kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/BatchAwareMemoryRecordsBuilder.java

```diff
@@ -0,0 +1,236 @@
+
+    /**
+     * Appends a record in the current batch
+     * @param record The record to append
+     * @return This builder
+     */
+    public @NonNull BatchAwareMemoryRecordsBuilder append(Record record) {
+        checkHasBatch();
+        builder.append(record);
+        return this;
+    }
+
+    /**
+     * Appends a record using a different offset in the current batch.
+     * @param offset The offset
+     * @param record The record to append
+     * @return This builder
+     */
+    public BatchAwareMemoryRecordsBuilder appendWithOffset(long offset, Record record) {
+        checkHasBatch();
+        builder.appendWithOffset(offset, record);
+        return this;
+    }
+
+    /**
+     * Append a new record at the given offset in the current batch.
+     * @param offset The absolute offset of the record in the log buffer
+     * @param timestamp The record timestamp
+     * @param key The record key
+     * @param value The record value
+     * @param headers The record headers if there are any
+     * @return This builder
+     */
+    public @NonNull BatchAwareMemoryRecordsBuilder appendWithOffset(long offset, long timestamp, byte[] key, byte[] value, Header[] headers) {
+        checkHasBatch();
+        builder.appendWithOffset(offset, timestamp, key, value, headers);
+        return this;
+    }
+
+    /**
+     * Append a new record at the given offset in the current batch.
+     * @param offset The absolute offset of the record in the log buffer
+     * @param timestamp The record timestamp
+     * @param key The record key
+     * @param value The record value
+     * @param headers The record headers if there are any
+     * @return This builder
+     */
+    public @NonNull BatchAwareMemoryRecordsBuilder appendWithOffset(long offset,
+                                                                    long timestamp,
+                                                                    ByteBuffer key,
+                                                                    ByteBuffer value,
+                                                                    Header[] headers) {
+        checkHasBatch();
+        builder.appendWithOffset(offset, timestamp, key, value, headers);
+        return this;
+    }
+
+    public @NonNull BatchAwareMemoryRecordsBuilder appendControlRecordWithOffset(long offset, @NonNull SimpleRecord record) {
+        checkHasBatch();
+        builder.appendControlRecordWithOffset(offset, record);
+        return this;
+    }
+
+    public @NonNull BatchAwareMemoryRecordsBuilder appendEndTxnMarker(long timestamp,
+                                                                      @NonNull EndTransactionMarker marker) {
+        checkHasBatch();
+        builder.appendEndTxnMarker(timestamp, marker);
+        return this;
+    }
+
+    /**
+     * Builds the memory records
+     * @return the memory records
+     */
+    public @NonNull MemoryRecords build() {
+        maybeAppendCurrentBatch();
```

Comment:
This differs from the semantics of MRB.build, which is idempotent.  Specifically it just returns `builtRecords` if it is already built.

I'm not sure if this is important to us right now, but I think the world will be simpler if we keep consistent behaviour.

Reply:
+1 on having it enter some closed state after the build.

Reply:
I can add some built-state management, but I don't think I can make it idempotent without some compromises. Here's my reasoning:

1. I want to allow a BAMRB to use a passed-in `ByteBufferOutputStream`. This allows the `ByteBufferOutputStream`'s buffer to be reused, minimising allocations of potentially large buffers. 
2. Thus BAMRB does not own the buffer that it returns, in particular it cannot guarantee that that buffer is not subsequently changes. Unknown to the first instance its `ByteBufferOutputStream` could be used with some other BAMRB instance such that calling `build()` a second time would return the same ByteBuffer but with different contents.

We can either stick to what's here currently (with some state flag to detect changes after `build()`, and some Javadoc), or sacrifice performance:
* either by making a BAMRB own its `ByteBufferOutputStream`
* or by taking a copy of the `ByteBuffer` in `build()`.

I'm inclined to preferring performance, because I think the potential for misuse is low.

Reply:
I added the kind of idempotency I think you meant @k-wall, where it's safe to call build() multiple times in the absence of modification of the buffer.